ios/LocalNetworkPrivacy/LocalNetworkPrivacy.m

```diff
@@ -26,22 +28,62 @@ - (void)dealloc {
 - (void)checkAccessState:(void (^)(BOOL))completion {
     self.completion = completion;
 
-    self.publishing = YES;
-    self.service.delegate = self;
-    [self.service publish];
+    BOOL permissionRequestedBefore = [[NSUserDefaults standardUserDefaults] boolForKey:kLocalNetworkPermissionRequestedKey];
 
-    self.timer = [NSTimer scheduledTimerWithTimeInterval:2 repeats:NO block:^(NSTimer * _Nonnull timer) {
-        [self.timer invalidate];
-        self.completion(NO);
-    }];
+    if (!permissionRequestedBefore) {
+        [self logPermissionStatus];
+        [[NSUserDefaults standardUserDefaults] setBool:YES forKey:kLocalNetworkPermissionRequestedKey];
+        [[NSUserDefaults standardUserDefaults] synchronize];
+
+        self.publishing = YES;
+        self.service.delegate = self;
+        [self.service publish];
+
+        self.timer = [NSTimer scheduledTimerWithTimeInterval:0.5 repeats:YES block:^(NSTimer * _Nonnull timer) {
+            if (self.publishing) {
+                NSLog(@"Local network permission status: Requesting");
+            } else {
+                [self.timer invalidate];
+                if (self.completion) {
+                    self.completion(self.service.includesPeerToPeer);
+                }
```

Comment:
create a separate function for completion

Reply:
done